fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerBlockPlaceEvents.java

```diff
@@ -0,0 +1,119 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.event.player;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public final class PlayerBlockPlaceEvents {
+	private PlayerBlockPlaceEvents() { }
+
+	/**
+	 * Callback before a block is placed.
+	 * Called on server only
+	 *
+	 * <p>If any listener cancels a block placing action, that block placing
+	 * action is canceled and {@link #CANCELED} event is fired. Otherwise, the
+	 * {@link #AFTER} event is fired.</p>
+	 */
+	public static final Event<Before> BEFORE = EventFactory.createArrayBacked(Before.class,
+			(listeners) -> (world, player, pos, state) -> {
+				for (Before event : listeners) {
+					boolean result = event.beforeBlockPlace(world, player, pos, state);
+
+					if (!result) {
+						return false;
+					}
+				}
+
+				return true;
+			}
+	);
+
+	/**
+	 * Callback after a block is placed.
+	 *
+	 * <p>Called on server only
+	 */
+	public static final Event<After> AFTER = EventFactory.createArrayBacked(After.class,
+			(listeners) -> (world, player, pos, state) -> {
+				for (After event : listeners) {
+					event.afterBlockPlace(world, player, pos, state);
+				}
+			}
+	);
+
+	/**
+	 * Callback when a block place has been canceled.
+	 *
+	 * <p>Called on server only. May be used on logical server to send packets to revert client-side block changes.
+	 */
+	public static final Event<Canceled> CANCELED = EventFactory.createArrayBacked(Canceled.class,
+			(listeners) -> (world, player, pos, state) -> {
+				for (Canceled event : listeners) {
+					event.onBlockPlaceCanceled(world, player, pos, state);
+				}
+			}
+	);
+
+	@FunctionalInterface
+	public interface Before {
+		/**
+		 * Called before a block is placed and allows canceling the block placing.
+		 *
+		 * <p>Implementations should not modify the world or assume the block place has completed or failed.</p>
+		 *
+		 * @param world the world in which the block is placed
+		 * @param player the player placing the block
+		 * @param pos the position at which the block is placed
+		 * @param state the block state <strong>before</strong> the block is placed
+		 * @return {@code false} to cancel block placing action, or {@code true} to pass to next listener
+		 */
+		boolean beforeBlockPlace(World world, PlayerEntity player, BlockPos pos, BlockState state);
```

Comment:
Since these events are only fired on the server, you should specify the more specific ServerWorld and ServerPlayerEntity as parameters.